core/system/src/mpu/vmpu.c

```diff
@@ -165,12 +165,76 @@ static void vmpu_sanity_check_box_namespace(int box_id, const char *const box_na
     } while (box_namespace[length]);
 }
 
+static void vmpu_box_index_init(uint8_t box_id, const uint32_t index_size, const UvisorBoxConfig *const config)
+{
+    void *box_bss;
+    UvisorBoxIndex *index;
+    uint32_t heap_size = config->heap_size;
+
+    if (box_id == 0)
+    {
+        if (!__uvisor_config.heap_start || !vmpu_sram_addr((uint32_t) __uvisor_config.heap_start)) {
+            HALT_ERROR(SANITY_CHECK_FAILED,
+                "Heap start pointer (0x%08x) is not in sram memory.\n",
+                (uint32_t) __uvisor_config.heap_start);
+        }
+        if (!__uvisor_config.heap_end || !vmpu_sram_addr((uint32_t) __uvisor_config.heap_end)) {
+            HALT_ERROR(SANITY_CHECK_FAILED,
+                "Heap end pointer (0x%08x) is not in sram memory.\n",
+                (uint32_t) __uvisor_config.heap_end);
+        }
+        if (__uvisor_config.heap_end < __uvisor_config.heap_start) {
+            HALT_ERROR(SANITY_CHECK_FAILED,
+                "Heap end pointer (0x%08x) is smaller than heap start pointer (0x%08x).\n",
+                (uint32_t) __uvisor_config.heap_end,
+                (uint32_t) __uvisor_config.heap_start);
+        }
+        /* box 0 still uses the main heap to be backwards compatible */
+        g_context_current_states[0].context = (uint32_t) __uvisor_config.heap_start;
+        heap_size = ((void *) __uvisor_config.heap_end -
+                     (void *) __uvisor_config.heap_start) -
+                    index_size;
+    }
+
+    box_bss = (void *) g_context_current_states[box_id].context;
+
+    /* The box index is at the beginning of the bss section */
+    index = box_bss;
+    /* Zero the _entire_ index, so that user data inside the box index is in a
+     * known state! This allows checking variables for `NULL`, or `0`, which
+     * indicates an initialization requirement. */
+    memset(box_bss, 0, index_size);
+    box_bss += index_size;
+    /* Initialize user context */
+    index->ctx = config->context_size ? box_bss : NULL;
+    box_bss += config->context_size;
+    /* Initialize process heap */
+    index->process_heap = heap_size ? box_bss : NULL;
+    index->process_heap_size = heap_size;
+    /* Active heap pointer is NULL, indicating that the process heap needs to
+     * be initialized on first malloc call! */
+    index->active_heap = NULL;
+
+    /* Cache the box id */
+    index->box_id = box_id;
+
+    /* point to the box config */
+    index->config = config;
+}
+
 static void vmpu_load_boxes(void)
 {
     int i, count;
     const UvisorBoxAclItem *region;
     const UvisorBoxConfig **box_cfgtbl;
     uint8_t box_id;
+    const uint32_t * const index_size = __uvisor_config.box_index_size;
```

Comment:
I'm wondering if we should read this from the box config instead of from `uvisor_config`, due to the risk of `uvisor_config` and box config not using the same value (perhaps by accident).